Create login system

radiantdinosaurs opened the pull request.

radiantdinosaurs committed:
Made a very simple application skeleton to get started.

radiantdinosaurs committed:
Added a .gitignore file.

radiantdinosaurs committed:
Made a very simple application skeleton to get started. (#12)

* Made a very simple application skeleton to get started.

* Added a .gitignore file.

radiantdinosaurs committed:
Created basic authentication.

Created a database for Slothbucket in MongoDB, as well as a 'users' collection. With this in place (and some other tools), users are able to register, login, view their profile, etc.
To achieve this, I made a lot of changes:
- Added in directories/files for routes, controllers, utils, and models
- Added in mongoose to simplify interactions with MongoDB
- Added in bcrypt for hashing and salting passwords
- Added in express-session and connect-mongo for handling and storing sessions
- Added in helmet for more secure HTTP headers
On top of this, I created the needed pages for basic routes in a user system (login, profile, and register). Also, I attempted to make the logging system more reusable by putting it in its own file and requiring it as needed, rather than having to type out all the winston/logger options every time.

andrewvora commented on ui/app.js:
```diff
@@ -0,0 +1,68 @@
+    // render the error page
+    response.status(error.status || 500)
+    response.render('error')
+})
```
This error handler should probably obfuscate all error information based on the environment. If it's `'development'` then it should give all info possible, but otherwise, the error message and stack trace should be hidden with some generic 404 page.

Also, doesn't seem like any of this gets called at all. At least, if you place a `console.log` sometime before `response.render('error')` and hit a nonexistent route like `/taco`. I would double check to see how to hand off the error from your middleware on L53

andrewvora commented on ui/app.js:
```diff
@@ -0,0 +1,68 @@
+    },
+    store: new MongoStore({
+        url: dbUrl
+    })
```
This is a suggestion for a different PR, but all configuration information like secrets and DB urls + credentials should be held in a config file or pulled from environment variables. If you keep it in a file, make sure that file is ignored and setup is detailed in a README local to the app.

andrewvora commented on ui/bin/www:
```diff
@@ -0,0 +1,48 @@
+    const port = parseInt(val, 10)
+    if (isNaN(port)) return val
+    if (port >= 0) return port
+    return false
```
Rather than return false, maybe just thrown an error. If the port isn't set up, it's gonna thrown an error down the line anyways.

andrewvora commented on ui/controllers/userController.js:
```diff
@@ -0,0 +1,109 @@
+'use strict'
```
Apparently, `use strict` is no longer necessary. So, I've read anyways. I have to remove it from my own projects, too

andrewvora commented on ui/controllers/userController.js:
```diff
@@ -0,0 +1,109 @@
+            } else return response.clearCookie('connect.sid', { path: '/' }).status(200).redirect('/')
+        })
+    } else response.status(200).redirect('/login')
+}
```
Filename should fall in line with the rest of the project.

`userController.js` -> `user-controller.js`

andrewvora commented on ui/controllers/userController.js:
```diff
@@ -0,0 +1,109 @@
+            })
+        }
+    }
+]
```
The final handler could be broken down into smaller functions. I would recommend breaking out the logic into another controller module. This should also make testing easier, instead of trying to wrestle out specific logic from this module

andrewvora commented on ui/controllers/userController.js:
```diff
@@ -0,0 +1,109 @@
+            })
+        }
+    }
+]
```
Same comment about this handler, break out logic into separate controller and include tests

andrewvora commented on ui/controllers/userController.js:
```diff
@@ -0,0 +1,109 @@
+            }
+            User.create(userData, (error, user) => {
+                if (error) {
+                    if (error.code === 11000) {
```
If this number was pulled from documentation somewhere, it would be better to define a top-level `const` at the top of this module and in the comments link the documentation that you pulled this value from

andrewvora commented on ui/models/user.js:
```diff
@@ -0,0 +1,75 @@
+'use strict'
```
Same note about `use strict` as earlier

andrewvora commented on ui/models/user.js:
```diff
@@ -0,0 +1,75 @@
+}
+
+const User = mongoose.model('Users', UserSchema)
+module.exports = User
```
Since there's validation logic in here, you should move the functions into their own module (like `mongoose-user-validation.js`) that exports the functions so you can use them in this `User` module. This way, you can test that logic without messing with Mongoose's magic

andrewvora replied:
Also, it'll come with the added benefit of moving some `require` dependencies out of this file and into the validation module

andrewvora reviewed (changes requested):
The logic itself seems sound, but there's lot of logic and no tests.

The bigger issue is that this kind of subverts the project direction by including DB and authentication logic into the UI app. The intent was to have all of that into the app module as additional routes and have UI only handle the display of data from the server app. Session management should be handled in the UI, but instead of accessing the database, you'll be hitting the server's API to get things like user objects and security tokens.

The "frontend" UI node app can do database operations, but nothing that's imperative to the slothbucket server. Keep it limited to UI-related operations.

**tl;dr** move all DB and auth logic into the `app/` dir and add tests. Continue managing session in the `ui/` dir.

radiantdinosaurs committed:
Finished working on request changes.

Created a config.js file to hold all sensitive credentials and added it to be ignored in the .gitignore files. Moved all the DB and authentication logic into the app/ directory and changed some code so it would all work. Continued managing session in the ui/ directory, storing the JWT/user id in the cookie and using MongoDB as the session store. Also, wrote some tests and did some other minor fixes.

radiantdinosaurs committed:
Added a mock for the config module to the unit test for verify.js

radiantdinosaurs committed:
Switched out config.js for env variables.

andrewvora commented on app/controllers/auth-controller.js:
```diff
@@ -0,0 +1,62 @@
+                password: request.body.password
+            }
+            userController.createUser(userData).then((user) => {
+                console.log(process.env.SECRET)
```
Instead of a just printing it out (which is a security risk) shouldn't you instead have handling logic? Something like `process.exit()` if it's not set? Since without it, your auth system shouldn't be functioning

andrewvora commented on app/app.js:
```diff
@@ -2,74 +2,40 @@
+const router = require('./routes/router')
+app.use('/', router)
+app.use((request, response, next) => next(returnError.resourceNotFound()))
+app.use((error, request, response, next) => {
```
These two handlers could use comments, so others wouldn't need to look up what they're used for

andrewvora commented on ui/app.js:
```diff
@@ -36,33 +34,29 @@ app.use(session({
+const router = require('./routes/router')
+app.use('/', router)
+app.use((request, response, next) => next(returnError.resourceNotFound()))
+app.use((error, request, response, next) => {
```
These two handlers could use comments describing their function

andrewvora commented on ui/controllers/auth-controller.js:
```diff
@@ -0,0 +1,90 @@
+            })
+        }
+    }
+]
```
The logic here could be broken out into smaller functions. Pass variables you need as params. It's difficult to read the logic without having to find out everything `post_register` does

If it makes it easier for you, separate the exports with multiline comments like

```javascript
/*------------------------*
 * POST /login
 *-------------------------*/
```

So the functions don't become hard to group

andrewvora commented on ui/controllers/auth-controller.js:
```diff
@@ -0,0 +1,90 @@
+                        request.session.jwt = body.token
+                        console.log(request.session)
+                        response.status(200).redirect('/')
+                    } else next(returnError.backendError())
```
Same here. Since the callback and if-elses already require so much nesting, any way you can flatten the structure would make it easier to read

andrewvora commented on app/controllers/user-controller.js:
```diff
@@ -0,0 +1,48 @@
+}
+
+module.exports.createUser = createUser
+module.exports.findUser = findUser
```
`user-controller` could use some tests

andrewvora commented on ui/controllers/auth-controller.js:
```diff
@@ -0,0 +1,90 @@
+            } else return response.clearCookie('connect.sid', { path: '/' }).status(200).redirect('/')
+        })
+    } else response.status(200).redirect('/login')
+}
```
`auth-controller` could use some tests

andrewvora reviewed (changes requested):
It's looking a lot better, but still has a lot of holes in terms of tests. There's a lot of code being added and, in the future, you want to be sure that upcoming changes aren't breaking anything vital. A broken test serves the purpose of telling a developer (hey, you broke something) and making them ask themselves, "did I mean to break this?"

radiantdinosaurs committed:
Fixed requested changes, added in tests, and did some minor restructuring and syntax fixes.

andrewvora reviewed (approved):
Let's get this merged in

radiantdinosaurs committed:
Merge remote-tracking branch 'origin/development' into create-login-system
